cachers/s3.go

```diff
@@ -0,0 +1,192 @@
+package cachers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"runtime"
+	"strings"
+
+	"github.com/aws/smithy-go"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+type S3Cache struct {
+	Bucket string
+	cfg    *aws.Config
+	// diskCache is where to write the output files to local disk, as required by the
+	// cache protocol.
+	diskCache *DiskCache
+
+	prefix string
+	// verbose optionally specifies whether to log verbose messages.
+	verbose bool
+
+	s3Client *s3.Client
+}
+
+func NewS3Cache(bucketName string, cfg *aws.Config, cacheKey string, disk *DiskCache, verbose bool) *S3Cache {
+	// get current architecture
+	arc := runtime.GOARCH
+	// get current operating system
+	os := runtime.GOOS
+	// get current version of Go
+	ver := strings.ReplaceAll(strings.ReplaceAll(runtime.Version(), " ", "-"), ":", "-")
+	prefix := fmt.Sprintf("cache/%s/%s/%s/%s", cacheKey, arc, os, ver)
+	log.Printf("S3Cache: configured to s3://%s/%s", bucketName, prefix)
+	return &S3Cache{
+		Bucket:    bucketName,
+		cfg:       cfg,
+		diskCache: disk,
+		prefix:    prefix,
+		verbose:   verbose,
+	}
+}
+
+func (c *S3Cache) client(ctx context.Context) (*s3.Client, error) {
+	if c.s3Client != nil {
+		return c.s3Client, nil
+	}
+	c.s3Client = s3.NewFromConfig(*c.cfg)
+	return c.s3Client, nil
+}
+
+func isNotFoundError(err error) bool {
+	if err != nil {
+		var ae smithy.APIError
+		if errors.As(err, &ae) {
+			code := ae.ErrorCode()
+			return code == "AccessDenied" || code == "NoSuchKey"
```

Comment:
Good question. Not sure why I chose AccessDenied, maybe my original permissions were only to `s3:GetObject` without `s3:ListBucket` which results in AccessDenied in case key cannot be found.